src/dataflow/src/source/mod.rs

```diff
@@ -184,23 +192,60 @@ lazy_static! {
         register_int_counter!("mz_bytes_read_total", "Count of bytes read from sources").unwrap();
 }
 
-/// Each source must implement this trait. Sources will then get created as part of the
-/// [`create_source`] function.
-pub trait SourceInfo {
-    /// Creates a new specialised SourceInfo object for a given source
+/// Creates a specific source, parameterised by 'Out'. Out denotes the encoding
+/// of the ingested data (Vec<u8> or Value).
+pub trait SourceConstructor<Out> {
+    /// Constructor for source creation
+    #[allow(clippy::too_many_arguments)]
     fn new(
         source_name: String,
         source_id: SourceInstanceId,
+        active: bool,
         worker_id: usize,
         worker_count: usize,
         consumer_activator: Arc<Mutex<SyncActivator>>,
         connector: ExternalSourceConnector,
+        consistency_info: &mut ConsistencyInfo,
+        encoding: DataEncoding,
     ) -> Self
     where
-        Self: Sized;
+        Self: Sized + SourceInfo<Out>;
+}
+
+/// Types that implement this trait expose a length function
+pub trait Length {
+    /// Returns the size of the object
+    fn len(&self) -> usize;
+    /// Returns true if the object is empty
+    fn is_empty(&self) -> bool;
+}
+
+impl Length for Vec<u8> {
+    fn len(&self) -> usize {
+        self.len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+impl Length for Value {
+    ///TODO(natacha): implement len() function for Value types
```

Comment:
Not really possible to compute a size in bytes without recursively traversing the entire tree.

Perhaps name this `MaybeLength`, have it return an `Option<usize>`, and document that it returns a size in bytes if such a thing is feasible to calculate?